src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -116,6 +116,8 @@ private sealed class Parser
 
             private readonly HashSet<TypeGenerationSpec> _implicitlyRegisteredTypes = new();
 
+            private readonly HashSet<ValueTuple<Type, DiagnosticDescriptor, string[]>> _typeLevelDiagnostics = new();
```

Comment:
Nit:
```suggestion
            private readonly HashSet<(Type, DiagnosticDescriptor, string[])> _typeLevelDiagnostics = new();
```
Could you also add a comment explaining what this tuple represents and what equality means here? Adding labels should help. Even better, it might make sense to factor this into a record type, which has the same brand of structural equality as ValueTuple but now you represent this type as a human-readable name.

Reply:
Added a comment. I've updated the collection type to list since we don't expect duplicates here.